src/command.ts

```diff
@@ -0,0 +1,236 @@
+      filePathFragments[baseDirIndex],
+    );
+    if (groupedFiles === undefined) {
+      groupedFiles = new Array();
+      filesGroupedPerDirectoryMap.set(filePathFragments[baseDirIndex], groupedFiles);
+    }
+    groupedFiles.push(path.join(...filePathFragments));
+  });
+
+  return [...filesGroupedPerDirectoryMap.values()];
+}
+
+(async () => {
+  let baseFiles: string[] | undefined;
+  let groupAmount: number;
+  const tokenizer = new CodeTokenizer(program.language);
+  const results: Map<string, Matches<number>> = new Map();
+  let baseFileMatches: Map<string, Matches<number>> = new Map();
+
+  // if the -d and the -b flag are active, filter out all the basefile locations
+  if (program.directory && program.base) {
+    program.base = path.normalize(program.base);
+
+    baseFiles = locations.filter(filePath => filePath.startsWith(program.base));
+    locations = locations.filter(filePath => !filePath.startsWith(program.base));
+  }
+
+  if (locations.length < 2) {
+    console.error("need at least two locations");
+    process.exit(1);
+  }
+
+  if (program.directory) {
+    if (program.base && baseFiles === undefined) {
+      console.error("no valid base files given");
+      process.exit(2);
+    }
+
+    const filesGroupPerDirectory: string[][] = groupPerDirectory(locations);
+
+    // If each program is a directory, then count the amount of directories.
+    groupAmount = filesGroupPerDirectory.length;
+
+    // If there is a base directory, compare all directories with it and put the results in one map.
+    if (program.base) {
+      for (const directory of filesGroupPerDirectory) {
+        const baseFileComparison = new Comparison(tokenizer);
+        await baseFileComparison.addFiles(baseFiles as string[]);
+        compose(
+          baseFileMatches,
+          await baseFileComparison.compareFiles(directory),
+        );
+      }
+    }
+
+    // Compare all directories with each other.
+    for (let i = 0; i < filesGroupPerDirectory.length; i += 1) {
+      for (let j = i + 1; j < filesGroupPerDirectory.length; j += 1) {
+        const comparison = new Comparison(tokenizer);
+        await comparison.addFiles(filesGroupPerDirectory[i]);
+        const matchesPerFile: Map<string, Matches<number>> = await comparison.compareFiles(
+          filesGroupPerDirectory[j],
+        );
+        compose(
+          results,
+          matchesPerFile,
+        );
+      }
+    }
+  } else {
+    // If each file is a separate program then count the amount of files.
+    groupAmount = locations.length;
+
+    // Compare the base file with all the other files when there is a base file.
+    if (program.base) {
+      const baseFileComparison = new Comparison(tokenizer);
+      await baseFileComparison.addFile(program.base);
+      baseFileMatches = await baseFileComparison.compareFiles(locations);
+    }
+
+    // Compare all the file with each other.
+    while (locations.length > 1) {
```

Comment:
This way, you're building the index for each file which is absolutely unnecessary. You only need to create it once.
- create a new comparison object
- add all files to the index (we have a utility method for this to do it for all files at once)
- for each of the files, query the index (we also have a utility function for this)

There is no need to aggregate the results yourself since this is already done by the comparison object.